### ISSUE TITLE
[models] Suggestion of GeneralizedRCNN forward output

### ISSUE THREAD

**Issue by frgfm** [pytorch/vision#1775]
## 🚀 Loss computation in eval mode

Even in `eval` mode, if the user passes targets to the `forward` method, return (or store) the loss dictionary in addition to the detections.

## Motivation

Recently, I made a quick training script for MaskRCNN on a different dataset. And since I usually look at training and validation loss evolution over epochs to spot potential overfit, I check how I could get around the current `forward` method of `GeneralizedRCNN`.

In train mode, it returns only the loss dictionary (which is fined, there is little use to check detections).
Now, say in eval mode that I want to get the loss. There is no way to retrieve it currently.

## Pitch

Change the `forward` method implementations of  `RoiHeads`, `RegionProposalNetwork`and ̀GeneralizedRCNN` so that when both an input and a target are passed to `GeneralizedRCNN.forward` it returns both the loss dictionary and the detections.

_Instead of checking  `self.training`, we could check `self.training or targets is not None`_

Happy to come up with a PR if you think that's a good idea, cheers!

**Comment by fmassa** [comment]
Hi,

Thanks for the feedback, I agree that this is something that should be improved.

I like your proposal, and I believe this is not a BC-breaking change, right?

In the future, I would go one step further: I think that we should always return `loss, detections`, in the model, at least some point in the future. This will be a BC-breaking change, so I'd like to think twice about it before committing to it.

Could you send a PR implementing what you proposed?

**Comment by frgfm** [comment]
Happy that you are onboard @fmassa!

I can come up with a PR next week! Just to be sure, for the time being, what are you advising to output with the `GeneralizedRCNN.forward` method if the condition is met?

I agree that both the `loss` and `detections` should be returned, and here without changing the number of outputs, even if the `RoiHeads` and `RegionProposalNetwork` compute and return the loss, they won't be returned. Concisely, I don't see the "non-breaking" option?

**Comment by frgfm** [comment]
Hi @fmassa,
Could you advise on what you had in mind for the non-breaking option?

**Comment by frgfm** [comment]
@fmassa let me know if you have some inputs on this
_If you think it's not worth it or irrelevant, we can close the issue_

**Comment by frgfm** [comment]
Very sorry, I thought I would have come up with a PR by now @fmassa @FrostByteGER

First, I do believe that slashing down the training time by default would be an issue. 

The current behaviour is the following (to the best of my knowledge):
training | targets | returned loss_dict | returned detections
-- | -- | -- | --
True | Y | Y | N
False | Y | N | Y
True | N | Not supported | Not supported
False | N | N | Y

Assuming, we potentially want to compute the loss_dict when passing targets, the behaviour needs to be updated in scenario 2. Additionally, if we want to get detections in training, the behaviour needs to be updated in scenario 1 as well.

Here is my suggestion: 
- new constructor flags: `return_loss_in_eval` and `return_detections_in_training` set to `False` by default
- which would give us: 

return_loss_in_eval | return_detections_in_training | training | targets | returned loss_dict | returned detections
-- | -- | -- | -- | -- | --
False| False | True | Y | Y | N
False| True | True | Y | Y | Y
False| False | False | Y | N | Y
True| False | False | Y | Y | Y
False| False | True | N | Not supported | Not supported
False| False | False | N | N | Y

What do you think? Suggestions for flag names are welcome :pray: 
Happy to work on a PR to handle this soon!

**Comment by frgfm** [comment]
Hey there :wave: 

I've seen that RCNN modifications are on the current roadmap of the library. Just checking whether you think the issue is still relevant @fmassa? 

Happy to try coming up with a PR or close the issue, but let's keep the number of open issues here rather low :)

**Comment by bencevans** [comment]
There's been further discussion more recently on #1574. Similar ideas have been drawn in terms of 2 arguments determining the return mode for each of loss and output, with #1574 introducing a 3rd overarching `return_tuples` with checks based on the return modes and targets availability given the values of the arguments.
